### PR TITLE
refactored bootstrap password

### PR DESCRIPTION
**Pull request by paynejacob**
https://github.com/rancher/rancher/issues/34686

# Problem
Occasionally when installing rancher the bootstrap password secret will not be set.

# Solution
I have not been able to consistently reproduce this issue but when I am able to reproduce it the cluster is under extreme load and errors about api timeouts are reported.  Likely the request to set the bootstrap password is failing but because we do not fail on this error rancher continues with creating the admin user loosing the bootstrap password forever.  To fix this I have made storing the bootstrap password a required prerequisite for creating the admin user.  This ensures that if the password is ever not saved the admin user will not be created and rancher will restart.  I also simplified the logic in the helm chart to avoid confusing situations where the bootstrap password is blank.

# Testing
- clean install with no given password
- clean install with given password
- docker run with no given password
- docker run with given password
- upgrade from 2.5.x
- upgrade from 2.5.x with the `--reuse-values` flag
- upgrade from 2.6.x
- run local

### CODE REVIEW

File: pkg/data/management/bootstrap_password.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+		generated = true
+		s.StringData[bootstrapPasswordSecretKey], _ = randomtoken.Generate()
+	}
+
+	// persist the password
+	if s.GetResourceVersion() != "" {
+		_, err = lister.Update(ctx, s, metav1.UpdateOptions{})
+	} else {
+		s.Name = bootstrapPasswordSecretName
+		_, err = lister.Create(ctx, s, metav1.CreateOptions{})
```
Comment (cmurphy) [inline pkg/data/management/bootstrap_password.go:50]: Is this needed? It wasn't in the original version, I'm not sure why not.

File: pkg/data/management/bootstrap_password.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+)
+
+const (
+	bootstrapPasswordSecretName = "bootstrap-secret"
+	bootstrapPasswordSecretKey  = "bootstrapPassword"
+)
+
+// GetBootstrapPassword reads the bootstrap password from it's secret.  If the secret is not found
+// it will be set from `CATTLE_BOOTSTRAP_PASSWORD` or generated if this is empty as well.
+func GetBootstrapPassword(ctx context.Context, lister corev1.SecretInterface) (string, bool, error) {
```
Comment (cmurphy) [inline pkg/data/management/bootstrap_password.go:21]: A `SecretLister` is a lister, `SecretInterface` is not, it's doing creates and updates.

**Review by cmurphy — COMMENTED** [review]
I can see what you meant about the original code being hard to follow :monocle_face:

File: pkg/data/management/bootstrap_password.go (hunk 1 above)
Comment (paynejacob) [reply pkg/data/management/bootstrap_password.go:50]: I believe we removed it from the original because of the user set it they should know it. I don't think there is any harm in storing it and it will help debugging in the future because we will know the secret should always get set for new installs.

File: pkg/data/management/bootstrap_password.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+	// load the existing secret
+	s, err = lister.Get(ctx, bootstrapPasswordSecretName, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return "", false, err
+	}
+
+	// if the bootstrap password is set return it
+	if s.StringData[bootstrapPasswordSecretKey] != "" {
+		return s.StringData[bootstrapPasswordSecretKey], generated, nil
```
[the remark this answers is not in the document]
Comment (paynejacob) [reply pkg/data/management/bootstrap_password.go:34]: No this code is only called once on the first start up when the admin user is created.  The bootstrap password is intended to only be used once to set a real password.

**Review by cmurphy — APPROVED** [review]

**Review by jakefhyde — APPROVED** [review]
LGTM
